src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs

```diff
@@ -476,29 +476,61 @@ private void AddCentralTransitiveDependencyGroupsForPackageReference(PackageSpec
 
             // Do not pack anything from the runtime graphs
             // The runtime graphs are added in addition to the graphs without a runtime
-            foreach (var targetGraph in targetGraphs.Where(targetGraph => string.IsNullOrEmpty(targetGraph.RuntimeIdentifier)))
+            foreach (RestoreTargetGraph targetGraph in targetGraphs.Where(targetGraph => string.IsNullOrEmpty(targetGraph.RuntimeIdentifier)))
             {
-                var centralPackageVersionsForFramework = project.TargetFrameworks.Where(tfmi => tfmi.FrameworkName.Equals(targetGraph.Framework)).FirstOrDefault()?.CentralPackageVersions;
+                TargetFrameworkInformation targetFrameworkInformation = project.TargetFrameworks.FirstOrDefault(i => i.FrameworkName.Equals(targetGraph.Framework));
+
+                if (targetFrameworkInformation == null)
+                {
+                    continue;
+                }
 
                 // The transitive dependencies enforced by the central package version management file are written to the assets to be used by the pack task.
                 IEnumerable<LibraryDependency> centralEnforcedTransitiveDependencies = targetGraph
-                    .Flattened
-                    .Where(graphItem => graphItem.IsCentralTransitive && centralPackageVersionsForFramework?.ContainsKey(graphItem.Key.Name) == true)
-                    .Select((graphItem) =>
+                    .Graphs
+                    .SelectMany(i => i.InnerNodes)
+                    .Where(node => node?.Item != null && node.Disposition == Disposition.Accepted && node.Item.IsCentralTransitive && targetFrameworkInformation.CentralPackageVersions?.ContainsKey(node.Item.Key.Name) == true)
+                    .Select((node) =>
```

Comment:
Can this lambda function be a static method, to ease debugging in the future?

```suggestion
                    .Select(node => CreateLibraryDependency(node))
```

Reply:
Unfortunately this lamda needs access to the instance member `_includeFlagGraphs` and local variables `project`, `targetGraph`, and `targetFrameworkInformation`.  Its not too bad to debug it by placing a breakpoint in the lamda.  The signature would look like this:

```c#
private static LibraryDependency GetLibraryDependencyForCentralTransitiveDependency(
    PackageSpec project,
    TargetFrameworkInformation targetFrameworkInformation,
    RestoreTargetGraph targetGraph,
    Dictionary<RestoreTargetGraph, Dictionary<string, LibraryIncludeFlags>> includeFlagGraphs,
    GraphNode<RemoteResolveResult> node)
```

What do you think works best here, lamda or static method?

Reply:
I personally prefer static methods because those encapsulate logic in a separate place. 

I think lambda as long as it does not modify any of the variables created outside of the lambda function. I just want to avoid any unwanted side effects.

Reply:
Actually just calling a static method at this point is still going to capture context and based on your comments I've realized that the real problem here is the complex LINQ statement.  So instead I'm going to wrap this whole bit of logic into a method that returns `IEnumerable<T>`.  Great suggestion!